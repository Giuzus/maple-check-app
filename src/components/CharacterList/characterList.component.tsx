import { useAtom } from "jotai";
import {
  charactersAtom,
  removeCharacterAtom,
} from "../../store/character.store";
import CharacterInfo from "../CharacterInfo/characterInfo.component";

function CharacterList() {
  const [characters] = useAtom(charactersAtom);
  const [, removeCharacter] = useAtom(removeCharacterAtom);
  return (
    <ul>
      {characters.map((char) => (
        <li key={char.name} className="flex p-4">
          <CharacterInfo character={char} />
          <button className="btn" onClick={() => removeCharacter(char.name)}>
            X
          </button>
        </li>
      ))}
    </ul>
  );
}

export default CharacterList;
