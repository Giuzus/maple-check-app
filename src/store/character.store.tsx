import axios from "axios";
import { atom } from "jotai";
import { atomWithStorage } from "jotai/utils";

export type Character = {
  name: string;
  error?: boolean;
  data?: {
    CharacterImageURL: string;
    Class: string;
    Level: number;
    Server: string;
  };
};

function addCharacter(characters: Character[], charName: string): Character[] {
  return [...characters, { name: charName }];
}

async function fetchCharacterData(
  characters: Character[],
  charName: string
): Promise<Character[]> {
  try {
    const response = await axios.get(
      `https://api.allorigins.win/raw?url=` +
        encodeURIComponent(
          `https://api.maplestory.gg/v2/public/character/gms/${charName}`
        )
    );

    if (response.data.status === "Error")
      throw new Error(response.data.message);

    return characters.map((char) =>
      char.name === charName
        ? { ...char, data: response.data.CharacterData }
        : char
    );
  } catch {
    console.error(`Error loading ${charName} info`);
    return characters.map((char) =>
      char.name === charName ? { ...char, error: true } : char
    );
  }
}

function removeCharacter(
  characters: Character[],
  charName: string
): Character[] {
  return characters.filter((char) => char.name !== charName);
}

export const charactersAtom = atomWithStorage<Character[]>("characters", []);
export const fetchCharacterDataAtom = atom(
  null,
  async (get, set, charName: string) => {
    set(
      charactersAtom,
      await fetchCharacterData(get(charactersAtom), charName)
    );
  }
);
export const addCharacterAtom = atom(
  null,
  async (get, set, charName: string) => {
    set(charactersAtom, addCharacter(get(charactersAtom), charName));
    set(fetchCharacterDataAtom, charName);
  }
);
export const removeCharacterAtom = atom(null, (get, set, charName: string) =>
  set(charactersAtom, removeCharacter(get(charactersAtom), charName))
);
