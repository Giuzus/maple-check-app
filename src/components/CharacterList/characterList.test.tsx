import { render, screen, within } from "@testing-library/react";
import { describe, expect, test } from "vitest";
import CharacterList from "./characterList.component";
import { charactersAtom } from "../../store/character.store";
import { HydrateAtoms } from "../../test/utils";
import userEvent from "@testing-library/user-event";
import { Provider } from "jotai";

const MOCK_CHARACTERS = [{ name: "char 1" }, { name: "char 2" }];

describe("Character list", () => {
  test("should match snapshot", () => {
    //Arrange
    const component = render(
      <Provider>
        <HydrateAtoms initialValues={[[charactersAtom, MOCK_CHARACTERS]]}>
          <CharacterList />
        </HydrateAtoms>
      </Provider>
    );

    //Assert
    expect(component.baseElement).toMatchSnapshot();
  });

  test("should remove characters", async () => {
    //Arrange
    const user = userEvent.setup();
    render(
      <Provider>
        <HydrateAtoms initialValues={[[charactersAtom, MOCK_CHARACTERS]]}>
          <CharacterList />
        </HydrateAtoms>
      </Provider>
    );
    const char1Del = within(
      screen.getByText("char 1").closest("li")!
    ).getByText("X");

    //Act
    await user.click(char1Del);

    //Assert
    expect(screen.queryByText("char 1")).toBeNull();
  });
});
