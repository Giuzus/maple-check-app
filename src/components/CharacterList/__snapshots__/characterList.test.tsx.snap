// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Character list > should match snapshot 1`] = `
<body>
  <div>
    <ul>
      <li
        class="flex p-4"
      >
        <div
          class="flex flex-col"
        >
          <span>
            char 1
          </span>
          <span>
            Loading...
          </span>
        </div>
        <button
          class="btn"
        >
          X
        </button>
      </li>
      <li
        class="flex p-4"
      >
        <div
          class="flex flex-col"
        >
          <span>
            char 2
          </span>
          <span>
            Loading...
          </span>
        </div>
        <button
          class="btn"
        >
          X
        </button>
      </li>
    </ul>
  </div>
</body>
`;
